cli/flox-rust-sdk/src/providers/container_builder.rs

```diff
@@ -0,0 +1,275 @@
+use std::io::{self, Write};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+use std::sync::LazyLock;
+
+use serde::Deserialize;
+use thiserror::Error;
+use tracing::debug;
+
+use super::buildenv::BuiltStorePath;
+use crate::providers::build::BUILDTIME_NIXPKGS_URL;
+use crate::providers::buildenv::NIX_BIN;
+use crate::utils::CommandExt;
+
+static MK_CONTAINER_NIX: LazyLock<PathBuf> = LazyLock::new(|| {
+    std::env::var("FLOX_MK_CONTAINER_NIX")
+        .unwrap_or_else(|_| env!("FLOX_MK_CONTAINER_NIX").to_string())
+        .into()
+});
+
+pub trait ContainerBuilder {
+    type Error: std::error::Error;
+    fn create_container_source(
+        &self,
+        name: impl AsRef<str>,
+        tag: impl AsRef<str>,
+    ) -> Result<ContainerSource, Self::Error>;
+}
+
+/// An implementation of [ContainerBuilder] that uses a Nix script
+/// to build a native [ContainerSource] from a [BuiltStorePath].
+///
+/// [MK_CONTAINER_NIX] is a Nix script that uses nixpkgs' `dockerTools.streamLayeredImage`
+/// which produces a script which that composes store paths
+/// as layers of a container image, and writes the resulting image to `stdout`.
+/// [MkContainerNix::create_container_source] performs the build of that script,
+/// which in turn builds all the required layers.
+/// The resulting [ContainerSource] will execute the script
+/// and stream the container image to a given sink via [ContainerSource::stream_container].
+///
+/// Note: that this implementation is not suitable for building containers on macOS.
+///
+/// 1. Containers are native _linux_ systems, for which packages can't be built on macOS.
+/// 2. The script generated by `dockerTools.streamLayeredImage` is not compatible with macOS.
+#[derive(Debug)]
+pub struct MkContainerNix {
+    store_path: BuiltStorePath,
+}
+
+#[derive(Debug, Error)]
+pub enum MkContainerNixError {
+    #[error("failed to call nix")]
+    CallNixError(#[source] std::io::Error),
+
+    #[error("failed to build container")]
+    BuildContainerError(String),
+
+    #[error("failed to parse nix build output")]
+    ParseBuildOutout(#[source] serde_json::Error),
+}
+
+impl MkContainerNix {
+    /// Create a new [MkContainerNix] instance that will build a container
+    /// from the given [BuiltStorePath].
+    /// Generally, this should be the output of a [crate::providers::buildenv::BuildEnv::build].
+    ///
+    /// Note: this constructor is only available on Linux.
+    /// On macOS, use a macOS-specific implementation of [ContainerBuilder].
+    #[cfg_attr(
+        not(target_os = "linux"),
+        deprecated(note = "MkContainerNix is not supported on this platform")
+    )]
```

Comment:
(non-blocking) Feels odd to call this deprecated when it was never functional and the cfg attrs feel about as fiddly as those that we decided against using in https://github.com/flox/flox/pull/2144. Seems like another compelling reason to have both implementations in the same provider.

Reply:
"deprecated" here was not meant as "discouraged from being used" (mainly due to it being non functional but alas ^^). Granted that meaning was cross checked with a dictionary and may not entirely fit into the theme of software, where it usually means "outdated".

iirc in #2144 the entire impl was basically disabled via cfg and eventually decided against because, while you can configure tooling to allow you to edit both, having two providers under the same name caused other funky warnings.

Here at most i'd cfg out the constructor, not the entire implementation, but deprecation seems to at least convey the intend without completely breaking compilation on either platform -- especially while we dont use platform dependent libraries.

> Seems like another compelling reason to have both implementations in the same provider.

Do you mean having both implementations of `ContainerBuilder` in the same module?
If so i'm not sure how that would be impacted by the previous point?

Reply:
🤷🏻  platform-specific deprecation still feels odd, but any alternative requires a smattering of more cfg's because the imports and setup are quite different for both code paths, so I'm fine to leave it.

> Do you mean having both implementations of `ContainerBuilder` in the same module?
> If so i'm not sure how that would be impacted by the previous point?

Yeah, because it would make it clearer that there are two platform specific implementations alongside each other.

I don't think it would be affected?

Reply:
> platform-specific deprecation still feels odd, but any alternative requires a smattering of more cfg's because the imports and setup are quite different for both code paths, so I'm fine to leave it.

the alternative is to implement `MkContainerNix::new` as with a `cfg`'d `panic!`/`unimplemented!`.
Its not quite the buildtime warning that you get from `deprecated` but still allows to compile it and prevents us from using it and getting bad errors later on from doing so.


```rust
    pub fn new(store_path: BuiltStorePath) -> Self {
        #[cfg(target_os = "linux")]
        Self { store_path }
		
		#[cfg(not(target_os = "linux"))]
		unimplemented!("MkContainerNix is not supported on this platform")
    }
```

Reply:
That's clearer 👍🏻

Reply:
> I don't think it would be affected?

Ok, the sentence sounded like the two are more closely related.

> Yeah, because it would make it clearer that there are two platform specific implementations alongside each other.

I'm not _strongly_ opinionated against putting the proxy implementation into the same module. I think due to using the _cli itself_ it creates an implicit dependency on the cli from the libs, but its still indirect via the docker container (i.e., this could be worse).

I think runtime detection and stuff is already separate from the `containerbuilder` implementation anyway. If you do feel strongly about this, we can move it in the other PR or any time later.

Reply:
I'd prefer we move it to the same module and I'll rebase #2478 after this is merged 👍🏻